addons/purchase_stock/wizard/purchase_order_suggest.py

```diff
@@ -202,3 +218,16 @@ def _save_values_for_vendor(self):
         condition = f'partner_id={self.partner_id.id}'
         for field in ['based_on', 'number_of_days', 'percent_factor']:
             self.env['ir.default'].set('purchase.order.suggest', field, self[field], company_id=cid, condition=condition)
+
+    def _get_on_demand_products_qty(self):
+        stock_moves = self.env['stock.move'].search([('date', '<=', fields.Datetime.now() + relativedelta(days=self.number_of_days)), ('state', '=', 'confirmed')])
+        product_qty_dict = {}
+
+        for stock_move in stock_moves:
+            product = stock_move.product_id
+            if product.id in product_qty_dict:
+                product_qty_dict[product.id] += stock_move.product_uom_qty
+            else:
+                product_qty_dict[product.id] = stock_move.product_uom_qty
+
+        return product_qty_dict
```

Comment:
You can use [`_read_group`](https://www.odoo.com/documentation/18.0/developer/reference/backend/orm.html#odoo.models.Model._read_group) for similar use cases.

```suggestion
qtys_by_product = self.env['stock.move']._read_group(
    domain=[('date', '<=', fields.Datetime.now() + relativedelta(days=self.number_of_days)), ('state', '=', 'confirmed')],
    groupby=['product_id'],
    aggregates=['product_uom_qty:sum']
)
return {product.id: product_demand for product, product_demand in qtys_by_product}
```